### ISSUE TITLE
Confusing error message if hostname cannot be found in configuration file

### ISSUE THREAD

**Issue by stevendanna** [chef/chef-server#49]
If the hostname of a given machine does not have a server configuration block in chef-server.rb, you will receive an error message like the following:

```
[root@backend1 opscode]# chef-server-ctl reconfigure
the ffi-yajl and yajl-ruby gems have incompatible C libyajl libs and should not be loaded in the same Ruby VM
falling back to ffi which might work (or might not, no promises)
ffi-yajl/json_gem is deprecated, these monkeypatches will be dropped shortly
the ffi-yajl and yajl-ruby gems have incompatible C libyajl libs and should not be loaded in the same Ruby VM
falling back to ffi which might work (or might not, no promises)
ffi-yajl/json_gem is deprecated, these monkeypatches will be dropped shortly
Starting Chef Client, version 11.12.2
Compiling Cookbooks...
Recipe: private-chef::default
* directory[/etc/opscode] action create (up to date)
* directory[/etc/opscode/logrotate.d] action create (up to date)

================================================================================
Recipe Compile Error in /opt/opscode/embedded/cookbooks/private-chef/recipes/default.rb
================================================================================


NoMethodError
-------------
undefined method `[]' for nil:NilClass


Cookbook Trace:
---------------
topology "ha"
/opt/opscode/embedded/cookbooks/private-chef/libraries/private_chef.rb:170:in `generate_secrets'
/opt/opscode/embedded/cookbooks/private-chef/libraries/private_chef.rb:471:in `generate_config'
/opt/opscode/embedded/cookbooks/private-chef/recipes/default.rb:87:in `from_file'


Relevant File Content:
----------------------
/opt/opscode/embedded/cookbooks/private-chef/libraries/private_chef.rb:

163: else
164: PrivateChef[k][pk] = p
165: end
166: end
167: end
168:
169: me = PrivateChef["servers"][node_name]
170>> ha_guard = PrivateChef['topology'] == 'ha' && !me['bootstrap']
171:
172: PrivateChef['rabbitmq']['password'] ||= generate_hex_if_bootstrap(50, ha_guard)
173: PrivateChef['rabbitmq']['jobs_password'] ||= generate_hex_if_bootstrap(50, ha_guard)
174: PrivateChef['rabbitmq']['actions_password'] ||= generate_hex_if_bootstrap(50, ha_guard)
175: PrivateChef['postgresql']['sql_password'] ||= generate_hex_if_bootstrap(50, ha_guard)
176: PrivateChef['postgresql']['sql_ro_password'] ||= generate_hex_if_bootstrap(50, ha_guard)
177: PrivateChef['drbd']['shared_secret'] ||= generate_hex_if_bootstrap(30, ha_guard)
178: PrivateChef['keepalived']['vrrp_instance_password'] ||= generate_hex_if_bootstrap(50, ha_guard)
179: PrivateChef['oc_bifrost']['superuser_id'] ||= generate_hex_if_bootstrap(16, ha_guard)




Running handlers:
[2014-12-19T17:23:39+00:00] ERROR: Running exception handlers
Running handlers complete

[2014-12-19T17:23:39+00:00] ERROR: Exception handlers complete
[2014-12-19T17:23:39+00:00] FATAL: Stacktrace dumped to /opt/opscode/embedded/cookbooks/cache/chef-stacktrace.out
Chef Client failed. 0 resources updated in 2.495912147 seconds
[2014-12-19T17:23:39+00:00] FATAL: NoMethodError: undefined method `[]' for nil:NilClass 
```

This is often the result of a typo.  A better error message would be:

```
ERROR: Cannot find configuration for this-host.example.com in /etc/opscode/chef-server.rb!  
```

Preferably without the large, unhelpful stack trace.

**Comment by xacaxulu** [comment]
+1

**Comment by brunobatista** [comment]
+1

**Comment by StephanMeijer** [comment]
+1

**Comment by sdelano** [comment]
This is a slightly different version of #21, but is all related to the same part of the code.

**Comment by JulianNL** [comment]
+1, for somebody messing around with Chef for the first time this is pretty unhelpfull

**Comment by stevendanna** [comment]
There are still many poor error messages that we might emit, but I think this particular one is solved.  Now, the error looks more like this:

```
[2016-11-03T00:53:24+00:00] FATAL: No server configuration found for "myhost" in /etc/opscode/chef-server.rb.
Server configuration exists for the following hostnames:

  some-other-server-1.chef.dev
  some-other-server-2.chef.dev

```

This was fixed in ef0d8b4aa43fa8f1366ee9d4ce4768a91f206b38
